### PR TITLE
Freeze and unfreeze view port

### LINKED ISSUE #1646: support freezing and unfreezing view port
### Description of Problem:
when a user clicks freeze on frontend, do not show new rows inserted to the table but still show updates of existing rows.
when a user clicks unfreeze, show all updates 

### Potential Solutions:
At backend, add a default column of creation time and last updated time to each table and filter the data based on creation time if the view port is froze and unfilter if view port is unfroze. 

The creation time is when the new row is added to the table.

### PR DESCRIPTION
**Pull request by tanjjj**
Default columns of creation time and last updated time are added to each table.
When a viewport is frozen, filter out the rows with creation time newer than freeze timestamp.
When a viewport is unfrozen, remove the filter.
If a viewport is already frozen, reject new freeze request.

And fixed an issue with RowPermissionFilter - user could remove any column from the viewport/vpColumns, hence we should give RowPermissionFilter the entire row of data

Note this PR doesn't cover freezing unfreezing view ports on a join table. Will do that in another PR for https://github.com/finos/vuu/issues/1653.

closes #1646

### CODE REVIEW

File: vuu/src/main/scala/org/finos/vuu/core/sort/FilterAndSort.scala [hunk 1]
```diff
@@ -90,6 +91,40 @@ case class RowPermissionFilter(checker: RowPermissionChecker) extends Filter wit
...
+    InMemTablePrimaryKeys(ImmutableArray.from[String](filtered))
+  }
+}
+
+case class RowPermissionAndFrozenTimeFilter(checker: RowPermissionChecker, frozenTime: Long) extends Filter with StrictLogging {
+  override def dofilter(source: RowSource, primaryKeys: TablePrimaryKeys, vpColumns: ViewPortColumns): TablePrimaryKeys = {
+    val filtered = primaryKeys.filter(key => {
+      try {
+        val rowData = source.pullRow(key)
+        checker.canSeeRow(rowData) && rowData.get(DefaultColumnNames.CreatedTimeColumnName).asInstanceOf[Long] < frozenTime
```
Comment (stonesmi) [inline vuu/src/main/scala/org/finos/vuu/core/sort/FilterAndSort.scala:116]: NIT: the Checker is likely to be slower than the the created time comparison, maybe flip the order here?

File: vuu/src/main/scala/org/finos/vuu/core/table/InMemDataTable.scala [hunk 2]
```diff
@@ -180,16 +180,19 @@ case class InMemDataTableData(data: ConcurrentHashMap[String, RowData], private
   protected def merge(update: RowData, data: RowData): RowData =
     MergeFunctions.mergeLeftToRight(update, data)
 
-  def update(key: String, update: RowData): TableData = {
+  def update(key: String, update: RowData, now: Long): TableData = {
 
     val table = data.synchronized {
```
Comment (stonesmi) [inline vuu/src/main/scala/org/finos/vuu/core/table/InMemDataTable.scala:186]: maybe the timprovider.now() call should be in this synchronized block, to guarantee ordering?

Also means you can back out the change to the TableData trait?

File: vuu/src/main/scala/org/finos/vuu/viewport/ViewPortContainer.scala [hunk 3]
```diff
@@ -317,6 +319,30 @@ class ViewPortContainer(val tableContainer: TableContainer, val providerContaine
...
+        }
+    }
+  }
+
+  def unfreezeViewPort(vpId: String): Unit = {
+    this.viewPorts.get(vpId) match {
+      case null =>
+        throw new Exception(s"Could not find viewport to unfreeze $vpId")
+      case vp: ViewPort =>
+        vp.unfreeze()
```
Comment (stonesmi) [inline vuu/src/main/scala/org/finos/vuu/viewport/ViewPortContainer.scala:341]: edge case for unfreezing a non-frozen VP?

File: vuu/src/test/scala/org/finos/vuu/core/sort/FilterAndSortFixture.scala [hunk 4]
```diff
@@ -3,12 +3,19 @@ package org.finos.vuu.core.sort
...
+import org.finos.toolbox.time.DefaultClock
 import org.finos.vuu.api.{Index, Indices, TableDef}
 import org.finos.vuu.core.filter.FilterClause
-import org.finos.vuu.core.table.{Columns, RowWithData, InMemDataTable, ViewPortColumnCreator}
+import org.finos.vuu.core.table.DefaultColumnNames.CreatedTimeColumnName
+import org.finos.vuu.core.table.{Columns, InMemDataTable, RowWithData, ViewPortColumnCreator}
 import org.finos.vuu.test.TestFriendlyJoinTableProvider
 
 object FilterAndSortFixture {
+  private val timeProvider = new DefaultClock
```
Comment (stonesmi) [inline vuu/src/test/scala/org/finos/vuu/core/sort/FilterAndSortFixture.scala:14]: there is a TestFriendlyClock class you can use here :)

**Review by stonesmi — CHANGES_REQUESTED** [review]

File: vuu/src/test/scala/org/finos/vuu/core/sort/FilterAndSortFixture.scala (hunk 4 above)
Comment (tanjjj) [reply vuu/src/test/scala/org/finos/vuu/core/sort/FilterAndSortFixture.scala]: nice! updated

File: vuu/src/main/scala/org/finos/vuu/viewport/ViewPortContainer.scala (hunk 3 above)
Comment (tanjjj) [reply vuu/src/main/scala/org/finos/vuu/viewport/ViewPortContainer.scala]: added a case to reject unfreeze request if vp not frozen

File: vuu/src/main/scala/org/finos/vuu/core/table/InMemDataTable.scala (hunk 2 above)
Comment (tanjjj) [reply vuu/src/main/scala/org/finos/vuu/core/table/InMemDataTable.scala]: updated

File: vuu/src/main/scala/org/finos/vuu/core/sort/FilterAndSort.scala (hunk 1 above)
Comment (tanjjj) [reply vuu/src/main/scala/org/finos/vuu/core/sort/FilterAndSort.scala]: updated
